### PR TITLE
[KIECLOUD-285] Enable basic upgrade support

### PR DESCRIPTION
**Pull request by tchughesiv**
also addresses -
https://issues.jboss.org/browse/KIECLOUD-287
https://issues.jboss.org/browse/KIECLOUD-288

Added initial CRD api conversion code. Can't fully enable until `CustomResourceWebhookConversion` feature is out of OpenShift Tech Preview.

Signed-off-by: tchughesiv <tchughesiv@gmail.com>

### CODE REVIEW

File: pkg/controller/kieapp/defaults/upgrade.go [hunk 1]
```diff
@@ -49,3 +52,70 @@ func MajorMinorMicro(productVersion string) (major, minor, micro string) {
 	}
 	return version[0], version[1], version[2]
 }
+
+// getConfigVersionDiffs ...
+func getConfigVersionDiffs(fromVersion, toVersion string, service api.PlatformService) (err error) {
```
Comment (bmozaffa) [inline pkg/controller/kieapp/defaults/upgrade.go]: This uses a named return variable, so you should just set the var and `return`... or don't name it and `return nil` or value as appropriate. It's confusing this way, though not incorrect as I can see implemented.

File: pkg/controller/kieapp/defaults/defaults.go [hunk 2]
```diff
@@ -31,10 +31,27 @@ var log = logs.GetLogger("kieapp.defaults")
...
 func GetEnvironment(cr *api.KieApp, service api.PlatformService) (api.Environment, error) {
+	minor, micro, err := checkProductUpgrade(cr)
+	if err != nil {
+		return api.Environment{}, err
+	}
+	// handle upgrade logic from here
+	minorVersion := getMinorImageVersion(cr.Spec.Version)
+	latestMinorVersion := getMinorImageVersion(constants.CurrentVersion)
+	if (micro && minorVersion == latestMinorVersion) ||
+		(minor && minorVersion != latestMinorVersion) {
```
[the remark this answers is not in the document]
Comment (tchughesiv) [reply pkg/controller/kieapp/defaults/defaults.go]: added a check for major

File: deploy/examples/v2/minor_upgrade.yaml [hunk 3]
```diff
@@ -5,4 +5,5 @@ metadata:
 spec:
   environment: rhpam-trial
   upgrades:
-    minor: true
+    enabled: true
+    minor: true
```
[the remark this answers is not in the document]
Comment (tchughesiv) [reply deploy/examples/v2/minor_upgrade.yaml]: added newline back

**Pull request merged**